core/src/main/scala/com/spingo/op_rabbit/ConnectionParams.scala

```diff
@@ -53,35 +55,133 @@ case class ConnectionParams(
+  private def fromParameters(config: Config): ConnectionParams = {
+    val hosts = readHosts(config).toArray
+    val port = config.getInt(PortPramPath)
+
+    ConnectionParams(
+      hosts = hosts.map { h => new Address(h, port) },
+      connectionTimeout = config.getDuration(ConnectionTimeoutParamPath, java.util.concurrent.TimeUnit.MILLISECONDS).toInt,
+      username = config.getString(UsernamePramPath),
+      password = config.getString(PasswordPramPath),
+      virtualHost = config.getString(VirtualHostPramPath),
+      ssl = config.getBoolean(SslPramPath)
+    )
+  }
+
+  private def fromUri(uri: URI): ConnectionParams = {
+    val uriAuthority = uri.getAuthority
+    val (hosts, (username, password)) = uriAuthority.indexOf('@') match {
+      case -1 =>
+        parseUriHosts(uriAuthority) -> parseUserPassword(None)
+      case idx =>
+        parseUriHosts(uriAuthority.substring(idx + 1)) -> parseUserPassword(Some(uriAuthority.substring(0, idx)))
+    }
+    val params = parseAndValidateUriQuery(Option(uri.getQuery))
+    ConnectionParams(
+      hosts = hosts,
+      username = username,
+      password = password,
+      virtualHost = Option(uri.getPath).fold(ConnectionFactory.DEFAULT_VHOST)(_.substring(1)),
+      ssl = TlsProtectedScheme == uri.getScheme,
+      connectionTimeout = params.connectionTimeout,
+      requestedHeartbeat = params.heartbeat,
+      requestedChannelMax = params.channelMax,
+      saslConfig = params.authMechanism
+    )
+  }
+
+  private def parseUriHosts(hosts: String): Seq[Address] = {
+    hosts.split(",").map(Address.parseAddress)
+  }
+
+  private def parseUserPassword(auth: Option[String]): (String, String) = {
+    val default = (ConnectionFactory.DEFAULT_USER, ConnectionFactory.DEFAULT_PASS)
+    auth.fold(default) { candidate =>
+      candidate.split(":").toList match {
+        case usr :: psswd :: _ => (usr, psswd)
+        case _ => default
+      }
+    }
+  }
+
   private def readHosts(config: Config): Seq[String] = {
-    Try(config.getStringList("hosts").asScala).getOrElse(readCommaSeparatedHosts(config))
+    Try(config.getStringList(HostsPramPath).asScala).getOrElse(readCommaSeparatedHosts(config))
   }
 
   private def readCommaSeparatedHosts(config: Config): Seq[String] =
     config
-      .getString("hosts")
+      .getString(HostsPramPath)
       .split(",")
       .map(_.trim)
 
-  def fromConfig(config: Config = RabbitConfig.connectionConfig) = {
-    val hosts = readHosts(config)
-    val port = config.getInt("port")
+  private def parseAndValidateUriQuery(query: Option[String]): UriQueryParams = {
+    query.fold(Array.empty[String])(_.split('&')).foldLeft(UriQueryParams()) {
+      case (resp, par) =>
+        val index = par.indexOf('=')
+        val key = par.substring(0, index)
+        val value = par.substring(index + 1)
+        key match {
+          case ConnectionTimeoutParamName =>
+            Try(resp.copy(connectionTimeout = value.toInt)).recover {
```

Comment:
Since `Try` is not used here in the monadic way, I think this is more readable
```scala
try resp.copy(connectionTimeout = value.toInt)
catch {
  case _: NumberFormatException =>
  throw new IllegalArgumentException("something...")
}
```

Reply:
In the core, I agree. My idea was to keep it compact, but it would be better to introduce a special function and use `try...catch` block.